File: src/solutions/easy/two-number-sum.js
// Best - O(n log (n)) time | O(1) space
function twoNumberSum(array, targetSum) {
  const sorted = array.sort((a,b) => a - b);
  let left = 0;
  let right = sorted.length - 1;

  while (left !== right) {
    let sum = sorted[left] + sorted[right];
    if (sum === targetSum) {
      return [sorted[left], sorted[right]];
    } else if (sum > targetSum) {
      right -= 1;
    } else if (sum < targetSum) {
      left += 1;
    }
  }
  return [];
}

// Incomplete
function twoNumberSum(array, targetSum) {
	for (let i = 0; i <= array.length; i += 1) {
		let firstNum = array[i];
		let secondNum = targetSum - firstNum;
		if (firstNum !== secondNum && array.includes(secondNum)) {
			return [firstNum, secondNum];
		}
	}
	
	return [];
}

// Worst - O(n^2) time | O(1) space
function twoNumberSum(array, targetSum) {
  for (let i = 0; i <= array.length; i += 1) {
    let compare = array[i];

    for (let k = 0; k <= array.length; k += 1) {
      let next = array[k+1];
      if (compare === next) {
        next = array[k+2];
      }

      let sum = compare + next;
      if (sum === targetSum) {
        return [compare, next];
      }
    }
  }

  return [];
}

File: javascript/solutions/easy/two_number_sum.go
package main

import (
	"fmt"
	"sort"
)

// O(n^2) time | O(1) space
func TwoNumberSum(array []int, target int) []int {
	var result = []int{}
	for x := 0; x < len(array); x++ {
		for y := x + 1; y < len(array); y++ {
			sum := array[x] + array[y]
			if sum == target {
				result = append(result, array[x], array[y])
			}
		}
	}

	return result
}

// O(n) time | O(n) space
func TwoNumberSum(array []int, target int) []int {
	nums := map[int]bool{}
	for _, num := range array {
		potentialMatch := target - num
		if _, found := nums[potentialMatch]; found {
			return []int{potentialMatch, num}
		} else {
			nums[num] = true
		}
	}
	return []int{}
}

// O(nlog(n)) time | O(1) space
func TwoNumberSum(array []int, target int) []int {
	sort.Ints(array)
	left, right := 0, len(array)-1
	for left < right {
		currentSum := array[left] + array[right]
		if currentSum == target {
			return []int{array[left], array[right]}
		} else if currentSum < target {
			left += 1
		} else {
			right -= 1
		}
	}
	return []int{}
}
